File: src/main/java/com/github/wgzhao/dbquery/controller/DBQueryController.java
package com.github.wgzhao.dbquery.controller;

import com.github.wgzhao.dbquery.dto.QueryResult;
import com.github.wgzhao.dbquery.entities.Sign;
import com.github.wgzhao.dbquery.errors.ParamException;
import com.github.wgzhao.dbquery.service.DBQueryService;
import com.github.wgzhao.dbquery.service.SignService;
import com.github.wgzhao.dbquery.util.ParamUtil;
import com.github.wgzhao.dbquery.util.SignUtil;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.env.Environment;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.sql.SQLException;
import java.util.List;
import java.util.Map;

import static com.github.wgzhao.dbquery.constant.Constants.APP_ID;
import static com.github.wgzhao.dbquery.constant.Constants.MAGIC_SIGN;
import static com.github.wgzhao.dbquery.constant.Constants.SIGN;


@RestController
@RequestMapping(value = "${app.api.query-prefix}")
@Slf4j
@Tag(name = "DBQuery", description = "DB Query APIs")
public class DBQueryController {

    private final DBQueryService queryService;

    private final SignService signService;

    private final Environment environment;

    public DBQueryController(DBQueryService queryService, SignService signService, Environment environment) {
        this.queryService = queryService;
        this.signService = signService;
        this.environment = environment;
    }

    private Map<String, String> queryParams;
    private Map<String, String> controlParams;
    private Map<String, String> allParams;

    private String errorMsg;

    private HttpServletResponse response;

    @GetMapping("/")
    public String index() {
        return "Hello, World!";
    }

    @GetMapping(value = "/query", produces = "application/json;charset=UTF-8")
    public QueryResult getQuery(@RequestParam() String selectId, @RequestParam Map<String, String> allParams,
                                    HttpServletRequest request, HttpServletResponse response) {
        this.response = response;
        this.allParams = allParams;
        return executeQuery(selectId);
    }

    @PostMapping("/query")
    public QueryResult postQuery(@RequestBody Map<String, String> payload, HttpServletRequest request, HttpServletResponse response) {
        String selectId = payload.getOrDefault("selectId", "");
        if (selectId.isEmpty()) {
            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
            return new QueryResult("缺少必要的参数 selectId");
        }
        this.response = response;
        this.allParams = payload;
        return executeQuery(selectId);
    }

    private QueryResult executeQuery(String selectId) {
        log.info("Begin to query with selectId {} and all params {}", selectId, allParams);
        //take the all params break into 2 parts, one is the query params, the other is the control params
        this.queryParams = ParamUtil.getQueryParams(allParams);
        this.controlParams = ParamUtil.getControlParams(allParams);

        if (!checkControlParams()) {
            log.error("Control params is invalid, error msg is {}", errorMsg);
            this.response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
            return new QueryResult(errorMsg);
        }

        // convert all query params key to lower case
        Map<String, String> lowerQueryParams = ParamUtil.lowercaseParams(queryParams);
        try {
            QueryResult queryResult = new QueryResult();
            List<Map<String, Object>> result = queryService.query(selectId, controlParams.get(APP_ID), lowerQueryParams);
            queryResult.setStatus(200);
            queryResult.setSuccess(true);
            queryResult.setMessage("success");
            queryResult.setTotal(result.size());
            queryResult.setData(Map.of("result", result));
            return queryResult;
        } catch (ParamException e) {
            log.error("invalid params: {}", e.getMessage());
            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
            return new QueryResult(e.getMessage());
        } catch (ClassNotFoundException | SQLException | RuntimeException e) {
            log.error("Query failed, error msg is {}", e.getMessage());
            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
            return new QueryResult(e.getMessage());
        }
    }

    /**
     * check the control params is valid or not
     *
     * @return true if valid, false if not
     */
    private boolean checkControlParams() {
        // check _sign has exists
        String sign = controlParams.getOrDefault(SIGN, null);
        if (sign == null || sign.isEmpty()) {
            this.errorMsg = "签名不存在";
            return false;
        }

        if (MAGIC_SIGN.equals(sign)) {
            // it's magic ,skip all validation
            final String[] activeProfiles = environment.getActiveProfiles();
            if (!(activeProfiles[0].equals("test") || activeProfiles[0].equals("dev"))) {
                this.errorMsg = "非法签名";
                return false;
            }
        } else {
            // check _appId is existing or not
            String appId = controlParams.getOrDefault(APP_ID, null);
            if (appId == null || appId.isEmpty()) {
                this.errorMsg = "缺少必要的参数 " + APP_ID;
                return false;
            }

            Sign signs = signService.querySign(appId);

            if (signs == null || signs.getAppKey() == null) {
                this.errorMsg = "无效的 " + APP_ID;
                return false;
            }

            if (!SignUtil.validSign(sign, queryParams, signs)) {
                this.errorMsg = "无效签名";
                return false;
            }
        }
        return true;
    }
}

File: src/main/kotlin/com/github/wgzhao/dbquery/controller/DBQueryController.kt
package com.github.wgzhao.dbquery.controller

import com.github.wgzhao.dbquery.constant.Constants
import com.github.wgzhao.dbquery.dto.QueryResult
import com.github.wgzhao.dbquery.service.DBQueryService
import com.github.wgzhao.dbquery.service.SignService
import com.github.wgzhao.dbquery.util.ParamUtil
import com.github.wgzhao.dbquery.util.SignUtil
import io.github.oshai.kotlinlogging.KotlinLogging
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.servlet.http.HttpServletResponse
import org.springframework.core.env.Environment
import org.springframework.util.Assert
import org.springframework.web.bind.annotation.*
import java.sql.SQLException

@RestController
@RequestMapping(value = ["\${app.api.query-prefix}"])
@Tag(name = "DBQuery", description = "DB Query APIs")
class DBQueryController(
    val queryService: DBQueryService,
    val signService: SignService,
    val environment: Environment
) {
    private val logger = KotlinLogging.logger {  }

    @GetMapping("/")
    fun index(): String {
        return "Hello, World!"
    }

    @GetMapping(value = ["/query"], produces = ["application/json;charset=UTF-8"])
    fun getQuery(
        @RequestParam selectId: String?,
        @RequestParam allParams: MutableMap<String, String>,
        response: HttpServletResponse
    ): QueryResult {
        return executeQuery(selectId, allParams, response)
    }

    @PostMapping("/query")
    fun postQuery(@RequestBody payload: MutableMap<String, String>, response: HttpServletResponse): QueryResult {
        val selectId = payload.getOrDefault("selectId", "")
        if (selectId.isEmpty()) {
            response.status = HttpServletResponse.SC_BAD_REQUEST
            return QueryResult.error(HttpServletResponse.SC_BAD_REQUEST, "缺少必要的参数 selectId")
        }
        return executeQuery(selectId, payload, response)
    }

    private fun executeQuery(
        selectId: String?,
        allParams: MutableMap<String, String>,
        response: HttpServletResponse
    ): QueryResult {
        logger.info("Begin to query with selectId {} and all params {}", selectId, allParams)

        //take the all params break into 2 parts, one is the query params, the other is the control params
        val queryParams = ParamUtil.getQueryParams(allParams)
        val controlParams = ParamUtil.getControlParams(allParams)

        try {
            checkControlParams(queryParams, controlParams)
        } catch (e: Exception) {
            logger.error {"Control params is invalid, error msg is ${e.message}" }
            response.status = HttpServletResponse.SC_BAD_REQUEST
            return QueryResult.error(HttpServletResponse.SC_BAD_REQUEST, e.message)
        }

        // convert all query params key to lower case
        val lowerQueryParams = ParamUtil.lowercaseParams(queryParams)
        try {
            // ensure selectId and appId non-null before calling service
            val sid = selectId ?: return QueryResult.error(HttpServletResponse.SC_BAD_REQUEST, "缺少必要的参数 selectId")
            val appId = controlParams[Constants.APP_ID]
                ?: return QueryResult.error(HttpServletResponse.SC_BAD_REQUEST, "缺少必要的参数 ${Constants.APP_ID}")

            // convert lowerQueryParams (MutableMap<String, String>) to MutableMap<String, String> expected by service
            val lowerQueryParamsGeneric: MutableMap<String, String> = HashMap()
            lowerQueryParams.forEach { (k, v) -> lowerQueryParamsGeneric[k] = v }

            val result = queryService.query(sid, appId, lowerQueryParamsGeneric.ifEmpty { null })
            // queryService may return nullable list with nullable maps; normalize to List<Map<String, Any>>?
            val normalizedResult: List<Map<String, Any>>? = result?.map { it }
            val queryResult = QueryResult.success(normalizedResult)
            return queryResult
        } catch (e: ClassNotFoundException) {
            logger.error {"Query failed, error msg is ${e.message}" }
            response.status = HttpServletResponse.SC_INTERNAL_SERVER_ERROR
            return QueryResult.error(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, e.message)
        } catch (e: SQLException) {
            logger.error {"Query failed, error msg is ${e.message}" }
            response.status = HttpServletResponse.SC_INTERNAL_SERVER_ERROR
            return QueryResult.error(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, e.message)
        } catch (e: RuntimeException) {
            logger.error {"Query failed, error msg is ${e.message}" }
            response.status = HttpServletResponse.SC_INTERNAL_SERVER_ERROR
            return QueryResult.error(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, e.message)
        }
    }

    /**
     * check the control params is valid or not
     */
    private fun checkControlParams(
        queryParams: MutableMap<String, String>?,
        controlParams: Map<String?, String?>
    ) {
        // check _sign has exists

        val sign = controlParams.getOrDefault(Constants.SIGN, null)
        Assert.hasLength(sign, "签名不存在")

        if (Constants.MAGIC_SIGN == sign) {
            // it's magic ,skip all validation
            val activeProfiles = environment.activeProfiles
            require(activeProfiles.isNotEmpty() && (activeProfiles[0] == "test" || activeProfiles[0] == "dev")) { "非法签名" }
        } else {
            // check _appId is existing or not
            val appId = controlParams.getOrDefault(Constants.APP_ID, null)
            Assert.hasLength(appId) { "缺少必要的参数 " + Constants.APP_ID }

            val signs = signService.querySign(appId!!)

            require(signs != null) { "无效的 " + Constants.APP_ID }

            require(SignUtil.validSign(sign, queryParams, signs)) { "无效签名 " }
        }
    }
}
